api/pkg/transformer/feast/transformer.go

```diff
@@ -337,62 +339,79 @@ func getExpressionExtractor(entity *transformer.Entity) string {
+func (t *Transformer) buildEntitiesRequest(ctx context.Context, request []byte, configEntities []*transformer.Entity, tableName string) ([]feast.Row, error) {
 	span, ctx := opentracing.StartSpanFromContext(ctx, "feast.buildEntitiesRequest")
+	span.SetTag("table.name", tableName)
 	defer span.Finish()
 
-	var entities []feast.Row
 	var nodesBody interface{}
 	err := json.Unmarshal(request, &nodesBody)
 	if err != nil {
 		return nil, err
 	}
 
-	for _, configEntity := range configEntities {
-		switch configEntity.Extractor.(type) {
-		case *transformer.Entity_JsonPath:
-			_, ok := t.compiledJsonPath[configEntity.GetJsonPath()]
-			if !ok {
-				c, err := jsonpath.Compile(configEntity.GetJsonPath())
-				if err != nil {
-					return nil, fmt.Errorf("unable to compile jsonpath for entity %s: %s", configEntity.Name, configEntity.GetJsonPath())
-				}
-				t.compiledJsonPath[configEntity.GetJsonPath()] = c
-			}
-		}
+	var allSeries [][]*types.Value
+	maxLength := 1
 
+	for k, configEntity := range configEntities {
 		expressionExtractor := getExpressionExtractor(configEntity)
+
 		vals, err := getValuesFromJSONPayload(nodesBody, configEntity, t.compiledJsonPath[configEntity.GetJsonPath()], t.compiledUdf[expressionExtractor])
 		if err != nil {
 			return nil, fmt.Errorf("unable to extract entity %s: %v", configEntity.Name, err)
 		}
 
-		if len(entities) == 0 {
-			for _, val := range vals {
-				entities = append(entities, feast.Row{
-					configEntity.Name: val,
-				})
-			}
-		} else {
-			newEntities := []feast.Row{}
-			for _, entity := range entities {
-				for _, val := range vals {
-					newFeastRow := feast.Row{}
-					for k, v := range entity {
-						newFeastRow[k] = v
-					}
+		seriesLength := len(vals)
 
-					newFeastRow[configEntity.Name] = val
-					newEntities = append(newEntities, newFeastRow)
-				}
-			}
-			entities = newEntities
+		if seriesLength != 1 && maxLength != 1 && seriesLength != maxLength {
+			return nil, fmt.Errorf("entity %s has different dimension", configEntities[k].Name)
+		}
+
+		if seriesLength > maxLength {
+			maxLength = seriesLength
+		}
+
+		allSeries = append(allSeries, vals)
+	}
+
+	entities := make([]feast.Row, maxLength)
+	for s, series := range allSeries {
+		entityName := configEntities[s].Name
+
+		if len(series) == 1 {
+			entities = t.broadcastSeries(entityName, series, entities)
+		} else {
+			entities = t.addSeries(entityName, series, entities)
 		}
 	}
 
 	return entities, nil
 }
 
+func (t *Transformer) broadcastSeries(entityName string, series []*types.Value, entities []feast.Row) []feast.Row {
+	for e := range entities {
+		newFeastRow := feast.Row{}
+		for k, v := range entities[e] {
+			newFeastRow[k] = v
+		}
+		newFeastRow[entityName] = series[0]
+		entities[e] = newFeastRow
+	}
+	return entities
+}
+
+func (t *Transformer) addSeries(entityName string, series []*types.Value, entities []feast.Row) []feast.Row {
```

Comment:
Same as https://github.com/gojek/merlin/pull/119/files#r618989906, it can be simplified to
```
	for idx, entity := range entities {
		entity[entityName] = series[idx]
	}
	return entities
```